lib/relation-definition.js

```diff
@@ -532,37 +532,85 @@ function lookupModel(models, modelName) {
+ * @param {Object|String} modelToRef Reference to Model object to which you are
+ *  creating the relation: model instance, model name, or name of relation to model.
+ * @param {Object} params relation params
+ * @param {Boolean} singularize Whether the modelToRef should be singularized when
+ *  looking-up modelTo
+ * @return {Object} modelTo Instance of the 'to' model
+ */
+function lookupModelTo(modelFrom, modelToRef, params, singularize) {
+  var modelTo;
+
+  if ('string' === typeof modelToRef) {
     var modelToName;
-    params.as = params.as || modelTo;
-    modelTo = params.model || modelTo;
-    if (typeof modelTo === 'string') {
-      modelToName = (singularize ? i8n.singularize(modelTo) : modelTo).toLowerCase();
-      modelTo = lookupModel(modelFrom.dataSource.modelBuilder.models, modelToName) || modelTo;
-    }
+    modelTo = params.model || modelToRef; // modelToRef is modelTo name
+    params.as = params.as || modelToRef;  // modelToRef is relationTo name
+
     if (typeof modelTo === 'string') {
-      modelToName = (singularize ? i8n.singularize(params.as) : params.as).toLowerCase();
-      modelTo = lookupModel(modelFrom.dataSource.modelBuilder.models, modelToName) || modelTo;
+      // lookup modelTo by name
+      modelToName = modelTo;
+      modelToName = (singularize ? i8n.singularize(modelToName) : modelToName).toLowerCase();
+      modelTo = lookupModel(modelFrom.dataSource.modelBuilder.models, modelToName);
     }
-    if (typeof modelTo !== 'function') {
-      throw new Error(g.f('Could not find "%s" relation for %s', params.as, modelFrom.modelName));
+    if (!modelTo) {
+      // lookup modelTo by relationTo name
+      var relationToName = params.as;
+      modelToName = (singularize ? i8n.singularize(relationToName) : relationToName).toLowerCase();
+      modelTo = lookupModel(modelFrom.dataSource.modelBuilder.models, modelToName);
     }
+  } else {
+    // modelToRef might already be an instance of model
+    modelTo = modelToRef;
+  }
+  if (typeof modelTo !== 'function') {
+    throw new Error(g.f('Could not find relation %s for model %s', params.as, modelFrom.modelName));
   }
   return modelTo;
 }
 
-/*!
+/*
  * Normalize polymorphic parameters
- * @param {Object|String} params Name of the polymorphic relation or params
+ * @param {Object|String|Boolean} polymorphic Param `polymorphic` of the relation.
+ * @param {String} relationName The name of the relation we are currently setting up.
  * @returns {Object} The normalized parameters
  */
-function polymorphicParams(params, as) {
-  if (typeof params === 'string') params = {as: params};
-  if (typeof params.as !== 'string') params.as = as || 'reference'; // default
-  params.foreignKey = params.foreignKey || i8n.camelize(params.as + '_id', true);
-  params.discriminator = params.discriminator || i8n.camelize(params.as + '_type', true);
-  return params;
+function polymorphicParams(polymorphic, relationName) {
```

Comment:
- `params` generally refers to root params of relation. `polymorphic` is used everywhere else.
- any reference to `polymorphic.as` keyword is removed so the 2nd parameter changes to `relationName`.